TDataFrame.hxx

```diff
@@ -856,182 +1070,172 @@ public:
 
    // dummy call, end of recursive chain of calls
-   bool CheckFilters(int, unsigned) {
-      return true;
-   }
+   bool CheckFilters(int, unsigned int) { return true; }
 
    // register a TActionResultPtr
-   void RegisterActionResult(TActionResultPtrBase* ptr) {
-      fActionResultsPtrs.emplace_back(ptr);
-   }
+   void RegisterActionResult(TActionResultPtrBase *ptr) { fActionResultsPtrs.emplace_back(ptr); }
 
-   ActionBaseVec fBookedActions;
-   FilterBaseVec fBookedFilters;
-   std::map<std::string, TmpBranchBasePtr> fBookedBranches;
-   TActionResultPtrBaseVec fActionResultsPtrs;
+   unsigned int GetNSlots() {return fNSlots;}
+
+   Internal::ActionBaseVec_t fBookedActions;
+   Details::FilterBaseVec_t fBookedFilters;
+   std::map<std::string, TmpBranchBasePtr_t> fBookedBranches;
+   std::vector<TActionResultPtrBase *> fActionResultsPtrs;
    std::string fTreeName;
-   TDirectory* fDirPtr = nullptr;
-   TTree* fTree = nullptr;
+   TDirectory *fDirPtr = nullptr;
+   TTree *fTree = nullptr;
    const BranchList fDefaultBranches;
    // always empty: each object in the chain copies this list from the previous
-   // and they must copy an empty list from the base TDataFrame
+   // and they must copy an empty list from the base TDataFrameImpl
    const BranchList fTmpBranches;
-   TDataFrame& fFirstData;
+   std::weak_ptr<TDataFrameImpl> fFirstData;
+   unsigned int fNSlots;
 };
 
+} // end NS Details
+
+} // end NS ROOT
 
-//********* FUNCTION AND METHOD DEFINITIONS *********//
-template<int S, typename T>
-T& GetBranchValue(TVBPtr& readerValue, unsigned slot, int entry, const std::string& branch, TDataFrame& df)
+// Functions and method implementations
+namespace ROOT {
+template <typename T>
+TDataFrameInterface<T>::TDataFrameInterface(const std::string &treeName, TDirectory *dirPtr,
+                                            const BranchList &defaultBranches)
+   : fProxiedPtr(std::make_shared<Details::TDataFrameImpl>(treeName, dirPtr, defaultBranches))
 {
-   if(readerValue == nullptr) {
-      // temporary branch
-      void* tmpBranchVal = df.GetTmpBranchValue(branch, slot, entry);
-      return *static_cast<T*>(tmpBranchVal);
-   } else {
-      // real branch
-      return **std::static_pointer_cast<TTreeReaderValue<T>>(readerValue);
-   }
+   fProxiedPtr->fFirstData = fProxiedPtr->shared_from_this();
 }
 
+template <typename T>
+TDataFrameInterface<T>::TDataFrameInterface(TTree &tree, const BranchList &defaultBranches)
+   : fProxiedPtr(std::make_shared<Details::TDataFrameImpl>(tree, defaultBranches))
+{
+   fProxiedPtr->fFirstData = fProxiedPtr->shared_from_this();
+}
 
 void TActionResultPtrBase::TriggerRun()
 {
-   fFirstData.Run();
+   fFirstData.lock()->Run();
 }
 
-
-TActionResultPtrBase::TActionResultPtrBase(TDataFrame& firstData) : fFirstData(firstData)
+TActionResultPtrBase::TActionResultPtrBase(std::weak_ptr<Details::TDataFrameImpl> firstData) : fFirstData(firstData)
```

Comment:
this constructor should probably take a `shared_ptr`, it does not make sense to try to build a `TActionResultPtr` with a potentially deleted `TDataFrameImpl`